Sources/Tracking/Background Queue/Queue.swift

```diff
@@ -1,20 +1,132 @@
+ A queue exists to make our public facing SDK functions synchronous:
+ `CustomerIO.shared.trackEvent("foo")` where we perform the API call sometime in the future
+ and handle errors/retry so customers don't have to.
+
+ The queue is designed with the main purpose of performing API calls. Let's show an example of
+ how it's recommended you use the queue and how *not* to use the queue.
+
+ Let's say you're identifying a profile.
+ ```
+ // recommended use of the background queue
+ func identifyProfile(identifier: String, data: Encodable) {
+    // perform all operations here, first before touching the queue
+    keyValueStorage.save(identifier)
+
+    if deviceTokenSavedToDifferentProfile {
+       // it's OK to add tasks to the queue here before we identify the new profile
+       queue.addTask(.deleteDeviceToken, identifier: oldProfileIdentifier)
+       keyValueStorage.delete(deviceToken)
+    }
+
+    // then, add a background queue task
+    queue.addTask(.identifyProfile, identifier: identifier)
+ }
+
+ // then later on in the code, the background queue task runs
+ // *all* of our logic for identifying a new profile.
+ func runQueueTask() {
+     httpClient.identifyProfile(newProfileIdentifier
+ }
+ ```
+
+ Not recommended way of using the background queue:
+ ```
+ func identifyProfile(identifier: String, data: Encodable) {
+    queue.addTask(.identifyProfile, identifier: identifier, oldProfileIdentifier)
+ }
+
+ // then later on in the code, the background queue task runs
+ // *all* of our logic for identifying a new profile.
+ func runQueueTask() {
+     let newProfileIdentifier = ...
+
+     keyValueStorage.save(identifier)
+
+     if deviceTokenSavedToDifferentProfile {
+        httpClient.delete(deviceTokenFromOldProfile)
+
+        keyValueStorage.delete(deviceToken)
+     }
+
+     httpClient.identifyProfile(newProfileIdentifier
+ }
+ ```
+ */
+public protocol Queue: AutoMockable {
+    /**
+     Add a task to the queue to be performed sometime in the future.
+
+     `data` - Probably a struct that contains "a snapshot" of the data needed to perform the
+     background task (probably a network request).
+     */
+    func addTask<TaskData: Codable>(
+        type: QueueTaskType,
+        // sourcery:Type=AnyEncodable
+        // sourcery:TypeCast="AnyEncodable(data)"
+        data: TaskData
+    ) -> (success: Bool, queueStatus: QueueStatus)
+    func run(onComplete: @escaping () -> Void)
 }
 
 // sourcery: InjectRegister = "Queue"
 public class CioQueue: Queue {
     private let storage: QueueStorage
     private let siteId: SiteId
+    private let runRequest: QueueRunRequest
+    private let jsonAdapter: JsonAdapter
+    private let logger: Logger
+    private let sdkConfigStore: SdkConfigStore
 
-    init(siteId: SiteId, storage: QueueStorage) {
+    init(
+        siteId: SiteId,
+        storage: QueueStorage,
+        runRequest: QueueRunRequest,
```

Comment:
Yes. 

I did try to add comments to each of the queue classes (queue, run request, runner, storage, manager) explaining the use case of each. 

Breaking the queue logic into separate files are for
1. readability of the code. encapsulation of logic. 
2. scalability (since the runner will grow more and more with each queue task type that we add)
3. memory safety. I try to limit our use of singletons so I encapsulated all of that into just the `QueueRequestManager` class to keep it as small as possible.

Reply:
found it a little hard to differentiate between them (re: readability) and understand their purpose. Of the three reasons for breaking, memory safety though is a trump card - assuming that doing it in a flat way would increase likelihood that we could become more of a mem :pig: - let's get a second opinion on the strategy here from @ami-aman 

re: keeping the runner as a pure http client, I think the question ends up being whether there are any things we've discussed adding to our SDK that would require different behaviour than that? I haven't seen anything, and would err on the side of handling our current known & planned needs, vs worrying too much about long term. If we need to do a major refactor in future to handle some fancier use case, we can definitely do that, not worth encoding complexity now for an unknown future

Reply:
Though there is nothing wrong with modulating the files into Manager, runner and handler. @levibostian has focussed on keeping the code segregated / modulated + clean considering the future of the SDK and also I believe, testability since testing gets easier and reliable with modulated code. I appreciate Levi's efforts in this.

But at this moment, as @hownowstephen said, when we are introducing Alphas and Betas and we are not even sure what exactly is the customer needs OR how would our releases be welcomed or accepted I think we should focus more on getting the work done and bringing out new features and understand what next would the customer want. As we introduce new features in our SDKs, code will be re-written sometimes or might (at least) need major/minor tweaks. So, we will have time to re-consider the code not modulated. And revisiting the code base later will also help us find out our own mistakes (or can call it as scope of improvement) which we, the developers usually miss out while development.

Reply:
I just took another glance at the files that make up the queue. Here they are: 

* `QueueRequestManager` - the small singleton
* `QueueRunner` - contains 1 function that gets called to run each of the queue tasks. 
* `QueueRunRequest` - The logic of the queue. 
* `QueueStorage` - wrapper around the file system to read/write data to files. 
* `Queue` - a delegate that is the entry point in the SDK to communicate with the SDK and all of it's features. 

Thinking about the above list, I don't see any value in combining any of the classes. Every scenario I have thought through in my head I am hesitant because of 1 of the 3 issues [I pointed out previously](https://github.com/customerio/customerio-ios/pull/78#discussion_r753383330). 

I can see something like combining `QueueRunRequest` and `QueueRunner` perhaps if we were able to think of a better replacement to the way `QueueRunner`.runTask()` works today (Maybe there is a way to reduce the copy/paste networking code that is in the runner today?). 

Are we on the same page or am I talking about something completely off?

Reply:
I'll defer to @ami-aman and possibly @Shahroz16 on the readability side, as I'm still quite new to Swift and am definitely not the target audience here!